packages/typescript-react-native-resolver/src/resolve.ts

```diff
@@ -0,0 +1,205 @@
+ * @param context Resolver context
+ * @param packageDir Root of the package which contains the module
+ * @param modulePath Optional relative path to the module
+ * @param extensions List of allowed file extensions, in order from highest precedence (index 0) to lowest.
+ * @returns Resolved module, or `undefined` if resolution fails
+ */
+export function resolveModule(
+  context: ResolverContext,
+  packageDir: string,
+  modulePath: string | undefined,
+  extensions: ts.Extension[]
+): ts.ResolvedModuleFull | undefined {
+  //  A module path was given. Use that to resolve the module to a file.
+  if (modulePath) {
+    return findModuleFile(context, packageDir, modulePath, extensions);
+  }
+
+  const { host, log } = context;
+
+  let module: ts.ResolvedModuleFull | undefined;
+
+  //  No path was given. Try resolving the module using package.json
+  //  properties.
+  log.log("Reading package.json from directory %o.", packageDir);
+  const result = host.readFile(path.join(packageDir, "package.json"));
+  if (!result) {
+    throw new Error(
+      `Failed to read package.json from directory '${packageDir}'`
+    );
+  }
+  const { types, typings, main } = JSON.parse(result);
+
+  //  Only consult 'types' and 'typings' properties when looking for
+  //  type files (.d.ts).
+  if (extensions.includes(ts.Extension.Dts)) {
+    if (typeof types === "string" && types.length > 0) {
+      log.log("Package has 'types' field '%s'.", types);
+      module = findModuleFile(context, packageDir, types, extensions);
+    } else if (typeof typings === "string" && typings.length > 0) {
+      log.log("Package has 'typings' field '%s'.", typings);
+      module = findModuleFile(context, packageDir, typings, extensions);
+    }
+  }
+  if (!module && typeof main === "string" && main.length > 0) {
+    log.log("Package has 'main' field '%s'.", main);
+    module = findModuleFile(context, packageDir, main, extensions);
+  }
+
+  //  Properties from package.json weren't able to resolve the module.
+  //  Try resolving it to an "index" file.
+  if (!module) {
+    log.log("Searching for index file.");
+    module = findModuleFile(context, packageDir, "index", extensions);
+  }
+
+  return module;
+}
+
+function findPackageDependencyDir(
+  context: ResolverContext,
+  ref: PackageModuleRef,
+  startDir: string
+): string | undefined {
+  const { log, host } = context;
+
+  log.log(
+    "Searching for external package %o starting in %o.",
+    ref.scope ? ref.scope + "/" + ref.name : ref.name,
+    startDir
+  );
+
+  const suffixDir = path.join("node_modules", ref.scope ?? "", ref.name);
+
+  let searchDir = startDir;
+  const { root: searchRoot } = path.parse(searchDir);
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const p = path.join(searchDir, suffixDir);
+    if (host.directoryExists(p)) {
+      return p;
+    }
+    if (searchDir === searchRoot) {
+      return undefined;
+    }
+    searchDir = path.dirname(searchDir);
+  }
```

Comment:
We can avoid the linter warning by comparing against root in while condition:
```suggestion
  while (searchDir !== searchRoot) {
    const p = path.join(searchDir, suffixDir);
    if (host.directoryExists(p)) {
      return p;
    }
    searchDir = path.dirname(searchDir);
  }
  return undefined;
```

Reply:
It's not quite the same thing. `while (true)` will test for existence at the root, where your changes will skip the root.